hal-core/src/addr.rs

```diff
@@ -0,0 +1,262 @@
+    }
+
+    /// Aligns `self` down to `align`.
+    ///
+    /// The specified alignment must be a power of two.
+    fn align_down<A: Into<usize>>(self, _align: A) -> Self {
+        unimplemented!("eliza")
+    }
+
+    /// Returns `true` if `self` is aligned on the specified alignment.
+    fn is_aligned<A: Into<usize>>(self, _align: A) -> bool {
+        unimplemented!("eliza")
+    }
+
+    fn as_ptr(&self) -> *const () {
+        unimplemented!("eliza")
+    }
+}
+
+impl PAddr {
+    #[cfg(target_pointer_width = "64")]
+    pub fn from_u64(u: u64) -> Self {
+        // TODO(eliza): ensure that this is a valid physical address?
+        PAddr(u as usize)
+    }
+}
+
+impl ops::Add<usize> for VAddr {
+    type Output = Self;
+    fn add(self, rhs: usize) -> Self {
+        VAddr(self.0 + rhs)
+    }
+}
+
+impl ops::Add for VAddr {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self {
+        VAddr(self.0 + rhs.0)
+    }
+}
+
+impl ops::AddAssign for VAddr {
+    fn add_assign(&mut self, rhs: Self) {
+        self.0 += rhs.0;
+    }
+}
+
+impl ops::AddAssign<usize> for VAddr {
+    fn add_assign(&mut self, rhs: usize) {
+        self.0 += rhs;
+    }
+}
+
+impl ops::Sub<usize> for VAddr {
+    type Output = Self;
+    fn sub(self, rhs: usize) -> Self {
+        VAddr(self.0 - rhs)
+    }
+}
+
+impl ops::Sub for VAddr {
+    type Output = Self;
+    fn sub(self, rhs: Self) -> Self {
+        VAddr(self.0 - rhs.0)
+    }
+}
+
+impl ops::SubAssign for VAddr {
+    fn sub_assign(&mut self, rhs: Self) {
+        self.0 -= rhs.0;
+    }
+}
+
+impl ops::SubAssign<usize> for VAddr {
+    fn sub_assign(&mut self, rhs: usize) {
+        self.0 -= rhs;
+    }
+}
+
+impl Address for VAddr {
+    fn as_usize(self) -> usize {
+        self.0 as usize
+    }
+
+    fn align_up<A: Into<usize>>(self, _align: A) -> Self {
+        unimplemented!("eliza")
+    }
+
+    /// Aligns `self` down to `align`.
+    ///
+    /// The specified alignment must be a power of two.
+    fn align_down<A: Into<usize>>(self, _align: A) -> Self {
+        unimplemented!("eliza")
+    }
+
+    /// Returns `true` if `self` is aligned on the specified alignment.
+    fn is_aligned<A: Into<usize>>(self, _align: A) -> bool {
+        unimplemented!("eliza")
+    }
+
+    fn as_ptr(&self) -> *const () {
+        unimplemented!("eliza")
+    }
+}
+
+impl VAddr {
+    #[cfg(target_pointer_width = "64")]
+    pub fn from_u64(u: u64) -> Self {
+        // TODO(eliza): ensure that this is a valid virtual address?
+        VAddr(u as usize)
+    }
+}
+
+impl fmt::Debug for VAddr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(width) = f.width() {
+            f.debug_tuple("VAddr")
+                .field(&format_args!("{:#0width$x}", self.0, width = width))
+                .finish()
+        } else {
+            f.debug_tuple("VAddr")
+                .field(&format_args!("{:#x}", self.0,))
+                .finish()
+        }
```

Comment:
I wish that you could pass an `Option<usize>` into `width = ` with `format_args!`. Would make stuff like this much nicer to read.

If we feel like being cheeky here, I think it's possible to write `fmt::LowerHex::fmt(&self.0, f)` to get the field printed, but we'd need to write out the `"VAddr(...)"` ourselves.

I could also imagine always wanting to use a full-width printout for our addresses based on the address width, which we could do with something like `format_args!("{:#0width$x}", self.0, width = mem::size_of::<usize>() * 2)`.

Reply:
i'm fine with changing the fmt impl, but i just snarfed it out of the hal_x86_64 crate and pasted it here...would prefer to make any behavioral or implementation changes separately.